### PR TITLE
Refactoring to allow for pollable timeout

### PR DESCRIPTION
**Pull request by tgross**
For https://github.com/joyent/containerpilot/issues/139

This PR lifts most of the existing logic from the work @justenwalker did in the `tasks` package so it can be reused to give us a single abstraction for running external commands for user-defined hooks for both blocking actions like the main app and co-processes (via `RunAndWait`), and non-blocking actions that should timeout (via `RunWithTimeout`).

So far I've ported the `services` package to use it. This is work in progress and I need to move some tests around and then continue the implementation into the other packages. From here it's mostly no new code though, just swapping out `exec.Cmd` for `command.Command` in places and deleting redundancies.

### CODE REVIEW

**Comment by tgross** [comment]
One of the little bonuses of this is that we create the `exec.Cmd` fresh from the parsed args when we make the external call only, so we parse args once but there's no funky logic with trying to recreate `exec.Cmd` in the right spot, which has been a source of bugs in the past.

**Comment by tgross** [comment]
Current build passes tests but:

> /cp/src/github.com/joyent/containerpilot/commands/args.go:48:1: comment on exported function ArgsToCommand should be of the form "ArgsToCommand ..."
/cp/src/github.com/joyent/containerpilot/commands/commands.go:13:6: exported type Command should have comment or be unexported
/cp/src/github.com/joyent/containerpilot/commands/commands.go:24:1: exported function NewCommand should have comment or be unexported
/cp/src/github.com/joyent/containerpilot/commands/commands.go:47:10: if block ends with a return statement, so drop this else and outdent its block (move short variable declaration to its own line if necessary)
Found 4 lint suggestions; failing.
make: *** [lint] Error 1
The command "make lint build" exited with 2.

Curses, foiled by the linter again!

**Comment by tgross** [comment]
This is in a working state now but it needs pass to clean up dead code and then some ergonomic tweaks in the `commands` package API.

There is one behavioral difference between this implementation and the existing, which is that debug logs are a lot more verbose. The verbosity we have for `tasks` will now be seen for all the other hooks, which should improve the supportability of ContainerPilot.

**Comment by tgross** [comment]
> some ergonomic tweaks in the commands package API,

Particularly, as I'm doing code cleanup here, I'm thinking it may have been a mistake to make `RunAndWait`, `RunAndWaitForOutput` and `RunWithTimeout` methods rather than functions because it means we've now got nil checks all over the code base instead of neatly wrapping them in the functions. I may undo that little bit here.

File: commands/commands.go [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+	"os/exec"
+	"syscall"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/joyent/containerpilot/utils"
+)
+
+// Command wraps an os/exec.Cmd with a timeout, logging, and arg parsing.
+type Command struct {
```
Comment (justenwalker) [inline commands/commands.go]: I'm really digging this new Command abstraction - good choice

**Comment by justenwalker** [comment]
> I may undo that little bit here because that's the sort of thing we will definitely miss in the future and create a bug.

Upon further inspection, I agree.  Much easier to do the nil checks inside a function call like `Run(command)` rather than `command.Run()` since you don't have to worry if command is nil in the former case.

**Comment by tgross** [comment]
> Much easier to do the nil checks inside a function call like Run(command) rather than command.Run() since you don't have to worry if command is nil in the former case every time you call it.

Yeah and I just found another place I missed it, too! Will do.

**Comment by tgross** [comment]
@justenwalker I've made that change. See what you think.

**Comment by tgross** [comment]
I've updated the PR title to reflect that this only lays the groundwork for us to have timeouts. It doesn't actually yet implement them except for the existing timeouts option in the `tasks` package. I'd like to do that work under separate PR.

**Comment by tgross** [comment]
@justenwalker did you have any more thoughts on this PR?

**Comment by justenwalker** [comment]
LGTM

**Pull request merged**